sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/data/EventData.scala

```diff
@@ -0,0 +1,154 @@
+ */
+
+package org.typelevel.otel4s.sdk
+package trace.data
+
+import cats.Hash
+import cats.Show
+import cats.syntax.show._
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.semconv.trace.attributes.SemanticAttributes
+
+import java.io.PrintWriter
+import java.io.StringWriter
+import scala.util.control.NonFatal
+
+/** Data representation of an event.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api/#add-events]]
+  */
+sealed trait EventData {
+
+  /** The name of the event.
+    */
+  def name: String
+
+  /** The epoch time in nanos of the event.
+    */
+  def epochNanos: Long
+
+  /** The attributes of the event.
+    */
+  def attributes: Attributes
+
+  override final def hashCode(): Int =
+    Hash[EventData].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: EventData => Hash[EventData].eqv(this, other)
+      case _                => false
+    }
+
+  override final def toString: String =
+    Show[EventData].show(this)
+}
+
+object EventData {
+  private val ExceptionEventName = "exception"
+
+  /** Creates [[EventData]] with the given arguments.
+    *
+    * @param name
+    *   the name of the event
+    *
+    * @param epochNanos
+    *   the epoch time in nanos of the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def apply(
+      name: String,
+      epochNanos: Long,
+      attributes: Attributes
+  ): EventData =
+    Impl(name, epochNanos, attributes)
+
+  /** Creates [[EventData]] from the given exception.
+    *
+    * The name of the even will be `exception`.
+    *
+    * Exception details (name, message, and stacktrace) will be added to the
+    * attributes.
+    *
+    * @param epochNanos
+    *   the epoch time in nanos of the event
+    *
+    * @param exception
+    *   the exception to associate with the event
+    *
+    * @param attributes
+    *   the attributes to associate with the event
+    */
+  def fromException(
+      epochNanos: Long,
+      exception: Throwable,
+      attributes: Attributes
+  ): EventData = {
+    val allAttributes = {
+      val builder = Vector.newBuilder[Attribute[_]]
+
+      builder.addOne(
+        Attribute(
+          SemanticAttributes.ExceptionType,
+          Option(exception.getClass.getName).getOrElse("")
+        )
+      )
+
+      val message = exception.getMessage
+      if (message != null) {
+        builder.addOne(Attribute(SemanticAttributes.ExceptionMessage, message))
+      }
+
+      val stringWriter = new StringWriter()
+
+      try {
+        val printWriter = new PrintWriter(stringWriter)
+        exception.printStackTrace(printWriter)
+      } catch {
+        case e if NonFatal(e) => ()
+      }
+
+      builder.addOne(
+        Attribute(SemanticAttributes.ExceptionStacktrace, stringWriter.toString)
+      )
+
+      builder.addAll(attributes.toList)
+
+      Attributes(builder.result(): _*)
```

Comment:
I feel like there's a lot of inefficient conversions here: `toList`, `.result()`, `: _*` ...

It can be a follow-up, but I wonder if we can either have a mutable `AttributesBuilder` that is backed by a `MapBuilder` or we could have `Attributes.unsafeFromMap` style builder.